backends/magma/grad_1d.cu

```diff
@@ -0,0 +1,213 @@
+        case  9: launch_failed = magma_grad_1d_ncomp<P, 9>(ncomp, dTgrad, transT, dU, u_elstride, u_compstride, dV, v_elstride, v_compstride, nelem, maxthreads, queue); break;
+        case 10: launch_failed = magma_grad_1d_ncomp<P,10>(ncomp, dTgrad, transT, dU, u_elstride, u_compstride, dV, v_elstride, v_compstride, nelem, maxthreads, queue); break;
+        default: launch_failed = 1;
+    }
+    return launch_failed;
+}
+
+
+//////////////////////////////////////////////////////////////////////////////////////////
+static magma_int_t 
+magma_grad_1d_ncomp_q_p(
+                magma_int_t P, magma_int_t Q, magma_int_t ncomp,
+                const CeedScalar *dTgrad, magma_trans_t transT,
+                const CeedScalar *dU, magma_int_t u_elstride, magma_int_t u_compstride, 
+                      CeedScalar *dV, magma_int_t v_elstride, magma_int_t v_compstride, 
+                magma_int_t nelem, magma_int_t maxthreads, magma_queue_t queue)
+{
+    magma_int_t launch_failed = 0;
+    switch(P) {
+        case  1: launch_failed = magma_grad_1d_ncomp_q< 1>(Q, ncomp, dTgrad, transT, dU, u_elstride, u_compstride, dV, v_elstride, v_compstride, nelem, maxthreads, queue); break;
+        case  2: launch_failed = magma_grad_1d_ncomp_q< 2>(Q, ncomp, dTgrad, transT, dU, u_elstride, u_compstride, dV, v_elstride, v_compstride, nelem, maxthreads, queue); break;
+        case  3: launch_failed = magma_grad_1d_ncomp_q< 3>(Q, ncomp, dTgrad, transT, dU, u_elstride, u_compstride, dV, v_elstride, v_compstride, nelem, maxthreads, queue); break;
+        case  4: launch_failed = magma_grad_1d_ncomp_q< 4>(Q, ncomp, dTgrad, transT, dU, u_elstride, u_compstride, dV, v_elstride, v_compstride, nelem, maxthreads, queue); break;
+        case  5: launch_failed = magma_grad_1d_ncomp_q< 5>(Q, ncomp, dTgrad, transT, dU, u_elstride, u_compstride, dV, v_elstride, v_compstride, nelem, maxthreads, queue); break;
+        case  6: launch_failed = magma_grad_1d_ncomp_q< 6>(Q, ncomp, dTgrad, transT, dU, u_elstride, u_compstride, dV, v_elstride, v_compstride, nelem, maxthreads, queue); break;
+        case  7: launch_failed = magma_grad_1d_ncomp_q< 7>(Q, ncomp, dTgrad, transT, dU, u_elstride, u_compstride, dV, v_elstride, v_compstride, nelem, maxthreads, queue); break;
+        case  8: launch_failed = magma_grad_1d_ncomp_q< 8>(Q, ncomp, dTgrad, transT, dU, u_elstride, u_compstride, dV, v_elstride, v_compstride, nelem, maxthreads, queue); break;
+        case  9: launch_failed = magma_grad_1d_ncomp_q< 9>(Q, ncomp, dTgrad, transT, dU, u_elstride, u_compstride, dV, v_elstride, v_compstride, nelem, maxthreads, queue); break;
+        case 10: launch_failed = magma_grad_1d_ncomp_q<10>(Q, ncomp, dTgrad, transT, dU, u_elstride, u_compstride, dV, v_elstride, v_compstride, nelem, maxthreads, queue); break;
+        default: launch_failed = 1;
```

Comment:
It'd be nice to eventually work out some sort of fallback or something. Right now `petsc-bpsraw` and `t506` fail (though we already knew this and its unrelated to this PR)

Reply:
There are two options here. One of them is to develop the fall-back kernel, which works regardless of the values of P and Q (which would be horribly slow BTW). The second is to move away from nvcc and use nvrtc instead (and eventually hiprtc For AMD GPUs). I prefer the second solution, but I have to discuss this internally with @stomov and @nbeams .

Reply:
2 would be nice. The current best performing CPU backend uses LIBXSMM to JiT kernels based upon values at runtime, and it seems to work well in our /cuda/gen backend as well.